TROUBLESHOOTING.md

```diff
@@ -1,7 +1,70 @@
-# Performance Caveats
+# Debugging
+
+Note that the information in this section is subject to be removed in future releases of
+the _PyTorch/XLA_ software, since many of them are peculiar to a given internal implementation
+which might change.
+
+Sometimes, bad things happen and a deeper look into the _PyTorch/XLA_ stack is necessary.
+The _PyTorch/XLA_ stack keeps a series of metrics and counters during its execution.
+The **first thing** to check when model is slow is to generate a metrics report.
+
+Metrics report is extremely helpful in diagonising issues. Please try to include it in your bug
+report sent to us if you have it.
+
+## Get A Metrics Report
+
+Put the following line in your program to generate a report:
+
+```Python
+print(torch_xla._XLAC._xla_metrics_report())
+```
+
+## Understand The Metrics Report
+
+The report includes things like:
+- how many time we issue _XLA_ compilations and time spent on issuing.
+- how many times we execute and time spent on execution
+- how many device data handles we create/destroy etc.
+
+This information is reported in terms of percentiles of the samples. An example is:
+
+```
+Metric: CompileTime
+  TotalSamples: 202
+  Counter: 06m09s401ms746.001us
+  ValueRate: 778ms572.062us / second
+  Rate: 0.425201 / second
+  Percentiles: 1%=001ms32.778us; 5%=001ms61.283us; 10%=001ms79.236us; 20%=001ms110.973us; 50%=001ms228.773us; 80%=001ms339.183us; 90%=001ms434.305us; 95%=002ms921.063us; 99%=21s102ms853.173us
+```
+
+We also provide counters, which are named integer variables which track internal software status. For example:
+
+```
+Counter: CachedSyncTensors
+  Value: 395
+```
+
+In this report, any counter that starts with `aten::`
+indicates a context switch between the XLA device and CPU, which can be a
+potential performance optimization area in the model code.
+
+Counters are useful to understand which operations are routed back to the CPU engine of _PyTorch_.
+They are fully qualified with their C++ namespace:
+
+```
+Counter: aten::nonzero
+  Value: 33
+```
+
+If you see `aten::` ops other than `nonzero` and `_local_scalar_dense`, that usually means a missing
+lowering in PyTorch/XLA. Feel free to open a feature request for it on [GitHub issues](https://github.com/pytorch/xla/issues).
+
+# Known Performance Caveats
 
 PyTorch/XLA behaves semantically like regular PyTorch and XLA tensors share the full tensor interface with CPU & GPU tensors.
-However, constraints in XLA/hardware and the lazy evaluation model suggest certain patterns might result in bad performance:
+However, constraints in XLA/hardware and the lazy evaluation model suggest certain patterns might result in bad performance.
+
+If you see your model suffers bad performance,
```

Comment:
I think this would be better as:

"If your model training shows bad performance, keep in mind the following caveats:"

Reply:
I would not bolt this to "training".
What if "inference" performance suffer?

Reply:
The model itself doesn't show bad performance, it's the training (or inference) which does. I thought the same but wanted to still be somewhat concise - maybe it's ok to leave "training" out then?

Reply:
The code never show bad performance, until you run it.
One could add "model execution" (or similar), but I think the original phrase would be understood by anybody.
Bolting it to _training_ seems wrong.